### ISSUE TITLE
Error in playbook "The task includes an option with an undefined variable. The error was: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'ansible_fqdn"

### ISSUE THREAD

**Issue by cranky-coder** [PyratLabs/ansible-role-k3s#131]
I am trying to deploy a pi4 cluster using your ansbile roles.  I have done this in past using this same role.  I am trying to set it up so that I have 3 groups in my inventory.  masters, workers, storagenodes.  3 Masters, 7 total nodes, 5 storagenodes.   Laid out like this.

masters:
  kube0, kube1, kube2
workers:
  (kube0, kube1, kube2) only because i haven't tainted them to not be workers
  kube3, kube4, kube5, kube6
storagenodes:
  kube0, kube1, kube2, kube5, kube6

below is my inventory file redacted of private info
```
all:
  hosts:
    kube-0:
      ansible_host: "192.168.2.40"
      new_hostname: "kube-0"
    kube-1:
      ansible_host: "192.168.2.41"
      new_hostname: "kube-1"
    kube-2:
      ansible_host: "192.168.2.42"
      new_hostname: "kube-2"
    kube-3:
      ansible_host: "192.168.2.43"
      new_hostname: "kube-3"
    kube-4:
      ansible_host: "192.168.2.44"
      new_hostname: "kube-4"
    kube-5:
      ansible_host: "192.168.2.45"
      new_hostname: "kube-5"
    kube-6:
      ansible_host: "192.168.2.46"
      new_hostname: "kube-6"

    
  vars:
    ansible_user: *****
    ansible_pass: ********
    ansible_python_interpreter: /usr/bin/python3
    ansible_ssh_pass: ******
    ansible_ssh_user: ********
    k3s_release_version: v1.19.7+k3s1
  children:
    masters:
      hosts:
        kube-0:
        kube-1:
      vars:
        cluster_vip: 192.168.2.60
        cluster_fqdn: api.crankycoder.net
        k3s_control_node: true
        k3s_server:
          - tls-san: 
            -  "{{ cluster_vip }}"
            -  "{{ cluster_fqdn }}"
          - disable:
            - "traefik"
            - "servicelb"
            - "metrics-server"
            - "local-storage"
            - "flannel"
    workers:
      vars:
        k3s_control_node: false
      hosts:
        kube-3:
        kube-4:
        kube-5:
        kube-6:
    storagepods:
      hosts:
        kube-0:
        kube-1:
        kube-2:
        kube-5:
        kube-6:
      vars:
        k3s_agent:
          - node-label:
            - "node.longhorn.io/create-default-disk=true"
    pis:
      hosts:
        kube-0:
        kube-1:
        kube-2:
        kube-3:
        kube-4:
        kube-5:
        kube-6:
        kube-7:
        kube-8:
        kube-9:
        kube-10:
        kube-11:
        kube-12:
        kube-13:
    k3s_cluster:
      children:
        masters:
        workers:
        storagepods:
```

As you can see i set control node on the masters to true.  workers set to false.  storage nodes only get a special label.  

However, when I run the playbook I keep getting this error.  

fatal: [kube-0]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: 'ansible.vars.hostvars.HostVarsVars object' has no attribute 'ansible_fqdn'\n\nThe error appears to be in 'roles/xanmanning.k3s/tasks/build/preconfigure-k3s.yml': line 64, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Ensure ansible_host is mapped to inventory_hostname\n  ^ here\n"}

I can't figure out why.  If I run a debug on the inventory for that same node.  "ansible_host": "192.168.2.40" is the ipaddress.  inventory_hostname="kube-0"

But the error is talking about ansible_fqdn.   

```
kube-0 | SUCCESS => {
    "hostvars[inventory_hostname]": {
        "ansible_check_mode": false,
        "ansible_diff_mode": false,
        "ansible_facts": {},
        "ansible_forks": 5,
        "ansible_host": "192.168.2.40",
        "ansible_inventory_sources": [
            "**********************"
        ],
        "ansible_pass": "**************",
        "ansible_playbook_python": "/usr/bin/python3",
        "ansible_python_interpreter": "/usr/bin/python3",
        "ansible_ssh_pass": "**********",
        "ansible_ssh_user": "**************",
        "ansible_user": "*****",
        "ansible_verbosity": 0,
        "ansible_version": {
            "full": "2.9.16",
            "major": 2,
            "minor": 9,
            "revision": 16,
            "string": "2.9.16"
        },
        "cluster_fqdn": "api.crankycoder.net",
        "cluster_vip": "192.168.2.60",
        "group_names": [
            "k3s_cluster",
            "masters",
            "pis",
            "storagepods"
        ],
        "groups": {
            "all": [
                "kube-0",
                "kube-1",
                "kube-3",
                "kube-4",
                "kube-5",
                "kube-6",
                "kube-2",
                "kube-7",
                "kube-8",
                "kube-9",
                "kube-10",
                "kube-11",
                "kube-12",
                "kube-13"
            ],
            "k3s_cluster": [
                "kube-0",
                "kube-1",
                "kube-3",
                "kube-4",
                "kube-5",
                "kube-6",
                "kube-2"
            ],
            "masters": [
                "kube-0",
                "kube-1"
            ],
            "pis": [
                "kube-0",
                "kube-1",
                "kube-2",
                "kube-3",
                "kube-4",
                "kube-5",
                "kube-6",
                "kube-7",
                "kube-8",
                "kube-9",
                "kube-10",
                "kube-11",
                "kube-12",
                "kube-13"
            ],
            "storagepods": [
                "kube-0",
                "kube-1",
                "kube-2",
                "kube-5",
                "kube-6"
            ],
            "ungrouped": [],
            "workers": [
                "kube-3",
                "kube-4",
                "kube-5",
                "kube-6"
            ]
        },
        "inventory_dir": "**************",
        "inventory_file": "**********",
        "inventory_hostname": "kube-0",
        "inventory_hostname_short": "kube-0",
        "k3s_agent": [
            {
                "node-label": [
                    "node.longhorn.io/create-default-disk=true"
                ]
            }
        ],
        "k3s_control_node": true,
        "k3s_release_version": "v1.19.7+k3s1",
        "k3s_server": [
            {
                "tls-san": [
                    "192.168.2.60",
                    "api.crankycoder.net"
                ]
            },
            {
                "disable": [
                    "traefik",
                    "servicelb",
                    "metrics-server",
                    "local-storage",
                    "flannel"
                ]
            }
        ],
        "new_hostname": "kube-0",
        "omit": "__omit_place_holder__7b0fb9f06bd85a4b4e9b0c24ed2f1f796642d65e",
        "playbook_dir": "**********************"
    }
}
```

**Comment by cranky-coder** [comment]
I have discovered what the problem is.  Some of these nodes aren't online as i was trying to bring the cluster up a little bit at a time.  Because of that, the ansible_fqdn wasn't even cached.  When I clear the inventory to only online machines it ran through no problem.

I am not an ansible expert, so I don't know for sure yet if this is possible, but for something like that task where it loops through the inventory, is there a way to tell it to only loop through inventory that connections have been established to?

Thanks!  cluster is up now!

**Comment by xanmanning** [comment]
I think I just need to replace `ansible_play_hosts_all` with `ansible_play_hosts`. I am sure there is a reason why I chose `ansible_play_hosts_all` but I can't remember off the top of my head.

https://github.com/PyratLabs/ansible-role-k3s/search?q=ansible_play_hosts_all

**Comment by cranky-coder** [comment]
I got around it.  If it's gets updated.  great.  Either way, I have been having FANTASTIC luck and work with your roles.  I am building out a 14 pi4 cluster with longhorn storage built-in using your roles.  VERY VERY happy so far!

**Comment by xanmanning** [comment]
Hopefully it will be fixed in v2.10.4 which will be available on Galaxy shortly.
